File: donater-bot/handlers/payment.js
import { Composer } from 'grammy';
import { config } from '../config.js';
import { orderCreate } from '../database/db.js';
import { registerOrder as rateRegisterOrder } from '../middlewares/antiSpam.js';
import { paymentDetailKeyboard, paymentMethodsKeyboard } from '../utils/keyboards.js';
import { paymentTitle, paymentNumber } from '../utils/formatter.js';
import { notifyAdminsNewOrder } from '../utils/notify.js';
import { sendMessageRaw } from '../utils/tg_api.js';

export const paymentComposer = new Composer();

paymentComposer.callbackQuery(/^pay:(dc_city|alif|mastercard|milli)$/, async (ctx) => {
  await ctx.answerCallbackQuery().catch(() => {});
  const chatId = ctx.chat?.id;
  if (!chatId) return;

  try {
    if (ctx.session.step !== 'choose_payment' || !ctx.session.draft) return;

    const method = ctx.match[1];
    ctx.session.draft.paymentMethod = method;
    ctx.session.step = 'await_receipt';

    const num = paymentNumber(method, config);
    const title = paymentTitle(method, ctx.i18n);
    const body = ctx.i18n('payment_details', {
      title,
      number: num,
      price: ctx.session.draft.priceTjs,
    });

    await sendMessageRaw(chatId, body, {
      parse_mode: 'HTML',
      reply_markup: paymentDetailKeyboard(num, ctx.i18n),
    });

    await ctx.reply(ctx.i18n('send_receipt'), { parse_mode: 'HTML' });
  } catch (e) {
    console.error(e);
    await ctx.reply(ctx.i18n('error_generic'));
  }
});

paymentComposer.callbackQuery('flow:back_payment', async (ctx) => {
  await ctx.answerCallbackQuery().catch(() => {});
  const chatId = ctx.chat?.id;
  if (!chatId) return;

  try {
    if (ctx.session.step !== 'await_receipt') return;
    ctx.session.step = 'choose_payment';
    await sendMessageRaw(chatId, ctx.i18n('choose_payment'), {
      parse_mode: 'HTML',
      reply_markup: paymentMethodsKeyboard(ctx.i18n),
    });
  } catch (e) {
    console.error(e);
  }
});

paymentComposer.on('message:photo', async (ctx, next) => {
  try {
    if (ctx.session.step !== 'await_receipt' || !ctx.session.draft) {
      return next();
    }

    const uid = ctx.from?.id;
    if (!uid) return next();

    const photos = ctx.message.photo;
    const fileId = photos[photos.length - 1].file_id;
    const d = ctx.session.draft;

    const orderId = orderCreate({
      userTgId: uid,
      username: ctx.from?.username || null,
      gameName: d.gameName,
      productLabel: d.productLabel,
      priceTjs: d.priceTjs,
      gameAccountId: d.accountId,
      paymentMethod: d.paymentMethod,
      receiptFileId: fileId,
    });

    rateRegisterOrder(uid);

    ctx.session.step = null;
    ctx.session.draft = null;

    await ctx.reply(ctx.i18n('order_accepted_user'), { parse_mode: 'HTML' });

    const orderRow = {
      id: orderId,
      user_tg_id: uid,
      username: ctx.from?.username || null,
      game_name: d.gameName,
      product_label: d.productLabel,
      price_tjs: d.priceTjs,
      game_account_id: d.accountId,
      payment_method: d.paymentMethod,
    };

    await notifyAdminsNewOrder(ctx.api, { ...orderRow, user_tg_id: uid }, fileId, ctx.i18n);
  } catch (e) {
    console.error(e);
    await ctx.reply(ctx.i18n('error_generic'));
  }
});

paymentComposer.on('message:text', async (ctx, next) => {
  try {
    if (ctx.session.step !== 'await_receipt') return next();
    await ctx.reply(ctx.i18n('need_photo'), { parse_mode: 'HTML' });
  } catch (e) {
    console.error(e);
  }
});

File: donater-bot/handlers/payment.py
import logging

from aiogram import F, Router
from aiogram.fsm.context import FSMContext
from aiogram.types import (
    CallbackQuery,
    InlineKeyboardButton,
    InlineKeyboardMarkup,
    Message,
)

from config import config
from database import queries
from middlewares.anti_spam import register_order_placed
from states.order import OrderFlow
from utils import tg_api
from utils.admin_notify import notify_new_order
from utils.formatter import format_money_display

log = logging.getLogger(__name__)
router = Router(name="payment")


def _pay_title(method: str, t) -> str:
    if method == "dc_city":
        return t("payment_dc")
    if method == "alif":
        return t("payment_alif")
    if method == "mastercard":
        return t("payment_mc")
    if method == "milli":
        return t("payment_milli")
    return method


def _pay_number(method: str) -> str:
    if method == "dc_city":
        return config.dc_city_number or "—"
    if method == "alif":
        return config.alif_number
    if method == "mastercard":
        return config.mastercard_number or "—"
    if method == "milli":
        return config.milli_number or "—"
    return "—"


def _payment_detail_kb(number: str, t):
    try:
        from aiogram.types import CopyTextButton

        row = [
            InlineKeyboardButton(
                text=t("copy_number"),
                copy_text=CopyTextButton(text=number.replace(" ", "")),
            )
        ]
    except Exception:
        row = [
            InlineKeyboardButton(
                text=t("copy_number"), callback_data=f"paycopy:{number[:60]}"
            )
        ]
    rows = [row]
    rows.append(
        [InlineKeyboardButton(text=t("confirm_no"), callback_data="flow:back_payment")]
    )
    return InlineKeyboardMarkup(inline_keyboard=rows)


@router.callback_query(F.data.regexp(r"^paycopy:(.+)$"))
async def cb_pay_copy(query: CallbackQuery, t):
    num = (query.data or "").split(":", 1)[-1]
    await query.answer(t("copy_fallback_alert", number=num), show_alert=True)


@router.callback_query(OrderFlow.payment_choice, F.data.regexp(r"^pay:(dc_city|alif|mastercard|milli)$"))
async def cb_pay_method(query: CallbackQuery, t, state: FSMContext):
    await query.answer()
    method = (query.data or "").split(":", 1)[-1]
    data = await state.get_data()
    draft = data.get("draft") or {}
    if not draft:
        return
    draft["payment_method"] = method
    await state.update_data(draft=draft, order_created=False)
    await state.set_state(OrderFlow.waiting_receipt)

    title = _pay_title(method, t)
    number = _pay_number(method)
    display = format_money_display(number)
    price = draft.get("price_tjs", 0)
    body = t("payment_card", title=title, number=display, price=price)
    kb = _payment_detail_kb(number, t)
    await query.message.answer(body, parse_mode="HTML", reply_markup=kb)
    await query.message.answer(t("send_receipt"), parse_mode="HTML")


@router.callback_query(OrderFlow.waiting_receipt, F.data == "flow:back_payment")
async def cb_back_payment(query: CallbackQuery, t, state: FSMContext):
    await query.answer()
    await state.set_state(OrderFlow.payment_choice)
    kb = tg_api.payment_methods_markup(
        t("payment_dc"),
        t("payment_alif"),
        t("payment_mc"),
        t("payment_milli"),
        t("confirm_no"),
    )
    await tg_api.raw_send_message(
        query.message.chat.id, t("choose_payment"), reply_markup=kb, parse_mode="HTML"
    )


@router.message(OrderFlow.waiting_receipt, F.photo)
async def msg_receipt_photo(message: Message, t, state: FSMContext):
    data = await state.get_data()
    if data.get("order_created"):
        return
    draft = data.get("draft") or {}
    if not draft:
        await state.clear()
        return
    uid = message.from_user.id
    photos = message.photo
    if not photos:
        return
    fid = photos[-1].file_id
    try:
        oid = await queries.order_create(
            uid,
            message.from_user.username,
            draft["game_name"],
            draft["product_label"],
            int(draft["price_tjs"]),
            str(draft["account_id"]),
            str(draft["payment_method"]),
            fid,
        )
        await state.update_data(order_created=True)
        register_order_placed(uid)
        await state.clear()
        await message.answer(
            t("order_accepted_user", id=oid),
            parse_mode="HTML",
        )
        await notify_new_order(
            message.bot,
            oid,
            uid,
            message.from_user.username,
            draft["game_name"],
            draft["product_label"],
            int(draft["price_tjs"]),
            str(draft["account_id"]),
            str(draft["payment_method"]),
            fid,
            t,
        )
    except Exception as exc:
        log.exception("order_create: %s", exc)
        await message.answer(t("error_generic"))


@router.message(OrderFlow.waiting_receipt, ~F.photo)
async def msg_receipt_wrong(message: Message, t):
    if message.text and message.text.startswith("/"):
        return
    await message.answer(t("need_photo"), parse_mode="HTML")
